python/mxnet/metric.py

```diff
@@ -646,9 +646,13 @@ def update(self, labels, preds):
         loss = 0.
         num = 0
         for label, pred in zip(labels, preds):
-            assert label.size == pred.size/pred.shape[-1], \
+            k = pred.shape[self.axis]
+            n = int(pred.size/k)
+            assert label.size == n, \
                 "shape mismatch: %s vs. %s"%(label.shape, pred.shape)
             label = label.as_in_context(pred.context).reshape((label.size,))
+            if len(pred.shape) > 2:
+                pred = pred.reshape((n, k))
```

Comment:
This wouldn't work if axis is not -1

Reply:
I also realized this when making the change. I will try to make some time today to look into it, but I will be OOTO for the next 10 days after.